File: Programmers/[42894] 블록 게임/woong-jae/42894.js
function solution(board) {
    let answer = 0;
    const N = board.length;
    
    const deleteBlock = (row, col, width, height) => {
        for(let r = row; r < row + height; r++) {
            for(let c = col; c < col + width; c++) {
                board[r][c] = 0;
            }
        }
    }
    
    const canFill = (row, col) => {
        for(let r = 0; r < row; r++) {
            if(board[r][col]) return false;
        }
        return true;
    }
    
    const canDelete = (row, col, width, height) => {
        if(row + height > N || col + width > N) return false;
        let emptyCount = 0, lastValue = -1;
        for(let r = row; r < row + height; r++) {
            for(let c = col; c < col + width; c++) {
                if(board[r][c] === 0) {
                    if(!canFill(r, c)) return false;
                    emptyCount++;
                    if(emptyCount > 2) return false;
                }
                else {
                    if(lastValue === -1) lastValue = board[r][c];
                    else if(lastValue !== board[r][c]) return false;
                }
            }
        }
        
        deleteBlock(row, col, width, height);
        
        return true;
    }
    
    while(1) {
        let deleted = 0;
        
        for(let row = 0; row < N; row++) {
            for(let col = 0; col < N; col++) {
                if(canDelete(row, col, 3, 2) || canDelete(row, col, 2, 3)) {
                    deleted++;
                }
            }
        }
        
        if(deleted === 0) break;
        answer += deleted;
    }
        
    return answer;
}

File: Programmers/[42894] 블록 게임/ONE/42894.java
class Solution {
    public int solution(int[][] board) {
        int answer = 0;
        int n = board.length;
        int count;

        do {
            count = 0;

            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++) {
                    if (i < n - 1 && j < n - 2 && find(i, j, 2, 3, board))
                        count++;
                    else if (i < n - 2 && j < n - 1 && find(i, j, 3, 2, board))
                        count++;
                }
            answer += count;

        } while (count != 0);

        return answer;
    }

    private boolean find(int row, int col, int h, int w, int[][] board) {
        int emptyCount = 0;
        int value = -1;

        for (int r = row; r < row + h; r++)
            for (int c = col; c < col + w; c++) {
                if (board[r][c] == 0) {
                    if (!canFill(r, c, board))
                        return false;
                    if (++emptyCount > 2)
                        return false;
                }
                else {
                    if(value != -1 && value != board[r][c])
                        return false;
                    value = board[r][c];
                }
            }

        erase(row, col, h, w, board);

        return true;
    }

    private boolean canFill(int r, int c, int[][] board) {
        for(int i = 0; i < r; i++)
            if(board[i][c] != 0)
                return false;
        return true;
    }

    private void erase(int row, int col, int h, int w, int[][] board) {
        for (int r = row; r < row + h; r++)
            for (int c = col; c < col + w; c++)
                board[r][c] = 0;
    }
}